### PR TITLE
feat: Optimize delete asset memory usage

### PR DESCRIPTION
**Pull request by sumslim**
As part of issue: https://github.com/goto/compass/issues/81
Optimizing memory usage for cancelFn is delete asset from ES using
**Cancel function within the context timeout:**
After the specified timeout, the context for the task automatically expires, and the cancellation function is invoked.

### CODE REVIEW

**Comment by sumslim** [comment]
Test case failing due to race condition detected, possibly due to an edge case:
We need to ensure that no operation modifies cancelFnList between appending a cancel function and calling removeCancelFnByIndex. Might need to use mutex to synchronize access to the cancelFnList, but this might bring contention overhead
cc @haveiss @irainia

File: core/asset/service.go [hunk 1]
```diff
@@ -44,7 +43,7 @@ type ServiceDeps struct {
 	Config        Config
 }
 
-func NewService(deps ServiceDeps) (service *Service, cancel func()) {
```
Comment (irainia) [inline core/asset/service.go:47]: please return the cancel func here. this is essential during shutdown process.

File: core/asset/service.go [hunk 2]
```diff
@@ -154,8 +148,10 @@ func (s *Service) DeleteAssets(ctx context.Context, request DeleteAssetsRequest)
 
 	if !request.DryRun && total > 0 {
 		newCtx, cancel := context.WithTimeout(context.Background(), s.config.DeleteAssetsTimeout)
-		s.cancelFnList = append(s.cancelFnList, cancel)
-		go s.executeDeleteAssets(newCtx, deleteSQLExpr)
+		go func() {
+			defer cancel()
```
Comment (irainia) [inline core/asset/service.go:152]: we don't really need to use `defer` here. just put the `cancel` after `s. executeDeleteAssets()` as they are just two lines

**Review by irainia — CHANGES_REQUESTED** [review]
Let me know if a separate discussion is required

File: core/asset/service.go (hunk 1 above)
Comment (sumslim) [reply core/asset/service.go:47]: there is no cancel func in the NewService function to return.
What should we return?

Comment (irainia) [reply core/asset/service.go:47]: exactly the point, please return the cancel function and its implementation for calling the cancel functions.

**Review by irainia — APPROVED** [review]
LGTM! Please don't forget to test. Approving it.
